### PR TITLE
Fix rehash functionality

### PR DESCRIPTION
**Pull request by MattAlp**
Working on #2263- `Hash#rehash` currently recalculates hashes, but does not remove duplicate as described in the spec. This PR addresses this by unlinking duplicate entries within a hash when a hash collision and equal values are detected.

### CODE REVIEW

**Comment by eregon** [comment]
Could you untag the spec and commit that change to show which spec now passes?
```
jt untag spec/ruby/core/hash/rehash_spec.rb 
```

File: src/main/java/org/truffleruby/core/hash/HashNodes.java [hunk 1]
```diff
@@ -900,14 +896,37 @@ protected RubyHash rehashBuckets(RubyHash hash,
...
+                            break;
+                        }
                         bucketEntry = bucketEntry.getNextInLookup();
                     }
+                    if (encounteredDuplicateKey || compareHashKeysNode.equalKeys(
+                            compareByIdentity,
+                            entry.getKey(),
+                            entry.getHashed(),
+                            bucketEntry.getKey(),
+                            bucketEntry.getHashed())) { // If the bucket contains a single entry, we never set the flag
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/hash/HashNodes.java:919]: Could you avoid this duplicated call by changing the loop above to be a `do/while` loop?

File: src/main/java/org/truffleruby/core/hash/HashNodes.java [hunk 2]
```diff
@@ -890,6 +885,7 @@ protected RubyHash rehashBuckets(RubyHash hash,
             Arrays.fill(entries, null);
 
             Entry entry = hash.firstInSequence;
+            Entry previousEntry = null;
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/hash/HashNodes.java:888]: `previousEntry` typically means the previous Entry in lookup, not in sequence, could you clarify by renaming it to `previousInSequence`?

File: src/main/java/org/truffleruby/core/hash/HashNodes.java [hunk 3]
```diff
@@ -900,14 +896,37 @@ protected RubyHash rehashBuckets(RubyHash hash,
...
                         bucketEntry = bucketEntry.getNextInLookup();
                     }
+                    if (encounteredDuplicateKey || compareHashKeysNode.equalKeys(
+                            compareByIdentity,
+                            entry.getKey(),
+                            entry.getHashed(),
+                            bucketEntry.getKey(),
+                            bucketEntry.getHashed())) { // If the bucket contains a single entry, we never set the flag
+                        if (previousEntry != null) {
+                            previousEntry.setNextInSequence(entry.getNextInSequence());
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/hash/HashNodes.java:921]: What about `entry.getNextInSequence().getPreviousInSequence()`, won't that still refer to `entry`?

File: CHANGELOG.md [hunk 4]
```diff
@@ -13,7 +13,7 @@ Bug fixes:
 * Fix `Thread.handle_interrupt` to defer non-pure interrupts until the end of the `handle_interrupt` block (#2219).
 * Clear and restore errinfo on entry and normal return from methods in C extensions (#2227).
 * Fix extra whitespace in squiggly heredoc with escaped newline (#2238, @wildmaples and @norswap).
-
+* Fix `Hash#rehash` to remove duplicate keys after modifications (#2266, @MattAlp)
 Compatibility:
```
Comment (chrisseaton) [inline CHANGELOG.md:17]: Broken empty line here.

File: src/main/java/org/truffleruby/core/hash/HashNodes.java [hunk 5]
```diff
@@ -866,14 +867,8 @@ protected RubyHash rehashPackedArray(RubyHash hash,
...
 
-            for (int n = 0; n < getLanguage().options.HASH_PACKED_ARRAY_MAX; n++) {
-                if (n < size) {
-                    PackedArrayStrategy.setHashed(
-                            store,
-                            n,
-                            hashNode.execute(PackedArrayStrategy.getKey(store, n), compareByIdentity));
-                }
-            }
+            PackedArrayStrategy.promoteToBuckets(getContext(), hash, store, size);
```
[the remark this answers is not in the document]
Comment (chrisseaton) [reply src/main/java/org/truffleruby/core/hash/HashNodes.java:870]: We weren't sure the best way to solve this - multiple nested loops? Could get pretty verbose when exploded? Or a temporary set to see if a value has already been included? Seemed like that was basically the bucket strategy anyway?

File: src/main/java/org/truffleruby/core/hash/HashNodes.java (hunk 3 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/hash/HashNodes.java:921]: I think we should use `BucketsStrategy.removeFromSequenceChain` here, then it should be fine.

File: src/main/java/org/truffleruby/core/hash/HashNodes.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (eregon) [reply src/main/java/org/truffleruby/core/hash/HashNodes.java:870]: A nested loop seems OK, there is already no `@ExplodeLoop` on this method anyway.
I think what's important is not so much the speed of `rehash` but that the resulting Hash still uses the more efficient packed array representation for other calls later on.

**Comment by bjfish** [comment]
> Could you untag the spec and commit that change to show which spec now passes?
> 
> ```
> jt untag spec/ruby/core/hash/rehash_spec.rb 
> ```

I think it would be nice if the spec was updated to also cover the bucket hash rehash if it's not currently covering it.

**Comment by chrisseaton** [comment]
We made progress on this today.

**Comment by chrisseaton** [comment]
@MattAlp do you have time to work on this soon? Not a problem if you don't! Don't feel pressured! But we'd probably have to move it on a bit ourselves if you don't soon.

**Comment by chrisseaton** [comment]
> I think it would be nice if the spec was updated to also cover the bucket hash rehash if it's not currently covering it.

I worry about baking in TruffleRuby implementation details to the spec.

Two better ideas:

1. put it in `spec/truffle`
2. rely on running the specs in a loop and using a _chaos_ node to randomise storage to fix this

**Comment by eregon** [comment]
> I worry about baking in TruffleRuby implementation details to the spec.

I wouldn't worry about that. Consider ruby/spec a test suite, not a specification.
From a test suite point of view most Ruby implementations have different storage strategies for small and large hashes, so one could just use e.g. a Hash with 10 elements to test that and mention in the spec description or a in comment it's specifically to test large hashes which might have a different representation.

**Pull request closed without merging**